samza-api/src/main/java/org/apache/samza/system/ExecutionEnvironment.java

```diff
@@ -70,4 +73,57 @@ static ExecutionEnvironment fromConfig(Config config) {
    */
   void run(StreamGraphBuilder graphBuilder, Config config);
 
+  /**
+   * Constructs a {@link StreamSpec} from the configuration for the specified streamId.
+   *
+   * The stream configurations are read from the following properties in the config:
+   * {@code streams.{$streamId}.*}
+   * <br>
+   * All properties matching this pattern are assumed to be system-specific with two exceptions. The following two
+   * properties are Samza properties which are used to bind the stream to a system and a physical resource on that system.
+   *
+   * <ul>
+   *   <li>samza.system -         The name of the System on which this stream will be used. If this property isn't defined
+   *                              the stream will be associated with the System defined in {@code job.default.system}</li>
+   *   <li>samza.physical.name -  The system-specific name for this stream. It could be a file URN, topic name, or other identifer.</li>
+   * </ul>
+   *
+   * @param streamId  The logical identifier for the stream in Samza.
+   * @return          The {@link StreamSpec} instance.
+   */
+  StreamSpec streamFromConfig(String streamId);
+
+  /**
+   * Constructs a {@link StreamSpec} from the configuration for the specified streamId.
+   *
+   * The stream configurations are read from the following properties in the config:
+   * {@code streams.{$streamId}.*}
+   * <br>
+   * All properties matching this pattern are assumed to be system-specific with one exception. The following
+   * property is a Samza property which is used to bind the stream to a system.
+   *
+   * <ul>
+   *   <li>samza.system - The name of the System on which this stream will be used. If this property isn't defined
+   *                      the stream will be associated with the System defined in {@code job.default.system}</li>
+   * </ul>
+   *
+   * @param streamId      The logical identifier for the stream in Samza.
+   * @param physicalName  The system-specific name for this stream. It could be a file URN, topic name, or other identifer.
+   * @return              The {@link StreamSpec} instance.
+   */
+  StreamSpec streamFromConfig(String streamId, String physicalName);
```

Comment:
Well, an application with 1 streamId could be deployed in multiple environments with different configs pointing to a different physicalName for each environment.
 
That's a 1:N relationship in terms of the values. But for any given config each stream has 1 streamId and 1 physicalName. So 1:1 in terms of the properties themselves.

A file URN is not a valid streamId because it would prevent us from properly parsing the config. But that same URN is totally valid as a physicalName. So I would argue that physicalName is a necessary attribute.

The real question here is where the physicalName should come from:
1. Exclusively from config. (I felt this would be tedious for users)
2. Same as streamId but overridable via config (this is the what the first signature of streamFromConfig() does)
3. Specified from code (this is what the other signatures do)

I think Prateek convinced me that we can start with signature 1 and expand to the others if necessary. Stay tuned for the next patch, which will do this.